### PR TITLE
Toolbar button visibility service

### PR DESCRIPTION
**Pull request by linusnorton**
- Move toolbar button visibility state into service and use DI 
- Move toolbar button state into ToolbarModule
- Nest toolbar components inside main toolbar

### CODE REVIEW

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  public showZoom = false;
+  public showRotate = false;
+  public showPresentationMode = false;
+  public showOpenFile = false;
+  public showBookmark = false;
+  public showHighlight = false;
+  public showSearchBarToggle = false;
+  public showSubToolbarToggle = false;
+  public showSidebarToggle = false;
+
```
Comment (russellwardink) [inline projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: Feels like the Annotation has been missed?

File: projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ export type HighlightMode = boolean;
 export type DrawMode = boolean;
 
 @Injectable({providedIn: 'root'})
```
Comment (russellwardink) [inline projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts]: As you’ve moved these as Providers in the Toolbar Module, providedIn: root is a bit conflicting, as it will no longer use the root injector. Might be worth removing.

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import { BehaviorSubject } from 'rxjs';
+
```
Comment (russellwardink) [inline projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: Missing the @Injectable annotation

File: tslint.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
       "severity": "warn"
     },
     "eofline": true,
-    "forin": true,
+    "forin": false,
```
Comment (russellwardink) [inline tslint.json]: why this change?

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts (hunk 1 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: These are the toolbar buttons rather than the toggles at the top of the page to turn features on/off

File: projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts (hunk 2 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts]: Are you sure? I thought it was natural for multiple classes to use the root injector... I will google it

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts (hunk 3 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: I noticed in the second PR :)

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts (hunk 1 above)
Comment (russellwardink) [reply projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: should the draw and highlight mode not be included then?

File: projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts (hunk 2 above)
Comment (russellwardink) [reply projects/media-viewer/src/lib/toolbar/toolbar-event.service.ts]: tbh, not 100% but i believe the providedIn is a short-cut… not sure which one would take precedence.

File: projects/media-viewer/src/lib/media-viewer.component.html [hunk 5]
```diff
@@ -30,8 +28,7 @@
                    [rotateOperation]="actionEvents?.rotate | async"
                    [setCurrentPage]="actionEvents?.setCurrentPage | async"
                    [changePageByDelta]="actionEvents?.changePageByDelta | async"
-                   [currentPageChanged]="currentPageChanged"
-                   [searchBarHidden]="toolbarButtonToggles?.searchBarHidden | async">
```
Comment (velimd) [inline projects/media-viewer/src/lib/media-viewer.component.html]: why was this removed?

File: tslint.json (hunk 4 above)
Comment (linusnorton) [reply tslint.json]: Long, long ago in a JavaScript far away... libraries would modify the prototype of javascript objects in order to paper over the shoddy standard library. For example, it's entirely possible to set `Object.prototype.foo = () => console.log('foo');` and every object will then have a foo function. Potentially this means using a for in can be dangerous as you end up iterating keys you didn't expect. This linting rule checks that you have a `.hasOwnProperty` on the key. I've disabled it as it's not something that is done any more.

File: projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts (hunk 1 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: yes, they should. I'm not sure we have added the ability to turn them on/off

Comment (russellwardink) [reply projects/media-viewer/src/lib/toolbar/toolbar-button-visibility.service.ts]: at present, i had them enabled under the wierd top toolbar Annotate switch, but they should in theory follow the same convention now outlined above

File: tslint.json (hunk 4 above)
Comment (russellwardink) [reply tslint.json]: okay cool

File: projects/media-viewer/src/lib/viewers/viewer-event.service.spec.ts [hunk 6]
```diff
@@ -1,12 +1,12 @@
-import {ViewerEventsService} from './viewer-events.service';
+import {ViewerEventService} from './viewer-event.service';
 import {Subscription} from 'rxjs';
```
Comment (velimd) [inline projects/media-viewer/src/lib/viewers/viewer-event.service.spec.ts]: spacing

File: projects/media-viewer/src/lib/media-viewer.component.html (hunk 5 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/media-viewer.component.html]: uses dependency injection now

File: projects/media-viewer/src/lib/media-viewer.component.html [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<div id="outerContainer" [ngClass]="{ sidebarOpen: toolbarButtonToggles?.sidebarOpen | async }">
+<div id="outerContainer" [ngClass]="{ sidebarOpen: toolbarButtons.sidebarOpen | async }">
```
[the remark this answers is not in the document]
Comment (linusnorton) [reply projects/media-viewer/src/lib/media-viewer.component.html]: nope

File: projects/media-viewer/src/lib/viewers/viewer-event.service.spec.ts (hunk 6 above)
Comment (linusnorton) [reply projects/media-viewer/src/lib/viewers/viewer-event.service.spec.ts]: I didn't want to make unnecessary formatting changes as part of this PR.... it's already big enough
